Add superenv for Linuxbrew

sjackman opened the pull request.

sjackman committed:
Add shims/linux/super

sjackman committed:
Add extend/os/linux/extend/ENV/super.rb

sjackman committed:
shims/linux/super/cc: Fix for Linuxbrew

sjackman committed:
cc: Remove -headerpad_max_install_names for Linux

sjackman commented:
@Linuxbrew/core As the title says. In this version, I've copied `Library/Homebrew/shims/linux/super/cc` from the Mac version and modified it just enough to get it up and running. It'll require much more work to bring it up to pace with `--env=std` in Linuxbrew.

sjackman committed:
Use --env=std by default for Linuxbrew

sjackman commented:
My last commit makes `--env=std` the default on Linuxbrew, so that `--env=super` is opt-in. With that change, we should be able to merge this PR now, and work on improving `superenv` together.

DoomHammer commented on Library/Homebrew/extend/ENV.rb:
```diff
@@ -4,7 +4,11 @@
 require "extend/ENV/super"
 
 def superenv?
-  ARGV.env != "std" && ARGV.env != "inherit" && Superenv.bin
```
Why remove `inherit`?

sjackman replied:
The behaviour hasn't changed. `--env=inherit` still works as before.

sjackman committed:
superenv: Set ld.so and RPATH for Linuxbrew

Add the opt_lib directory of each dependency to the RPATH after
$HOMEBREW_PREFIX/lib, which is searched first, allowing dependent
libraries to be found even when the keg is unlinked.

sjackman commented:
Succeeded on Linux. Unrelated (I think) error on Mac:
```
Error:
Satisfy Dependencies and Requirements::depends_on macos#test_0001_understands depends_on macos: <array>:
Hbc::CaskError: It seems there is already an App at '/tmp/homebrew-tests-20170215-10426-1fzztn2/prefix/appdir-1487198972-634/Caffeine.app'.
```
https://travis-ci.org/Linuxbrew/brew/jobs/202017165#L469

bw-yuhogg commented on Library/Homebrew/shims/linux/super/cc:
```diff
@@ -261,8 +262,12 @@ class Cmd
       args << "-no_weak_imports" if no_weak_imports?
     when :ccld, :cxxld
+      args << "-Wl,--dynamic-linker=#{prefix}/lib/ld.so"
+      args << "-Wl,-rpath," + rpath_paths
```
Does it make sense to skip this if rpath_paths is empty?

sjackman replied:
Yes. It shouldn't happen I think, but better to be paranoid. Thanks for catching that case. Fixed.

bw-yuhogg commented on Library/Homebrew/extend/os/linux/extend/ENV/super.rb:
```diff
@@ -3,4 +3,11 @@ module Superenv
+
+  def determine_rpath_paths
+    paths = ["#{HOMEBREW_PREFIX}/lib"]
+    paths += deps.map { |d| d.opt_lib.to_s }
```
Should this be restricted to runtime dependencies?

For `shellinabox` I got:

```bash
patchelf --print-rpath ~/.linuxbrew/bin/shellinaboxd 
/home/linuxbrew/.linuxbrew/lib:/home/linuxbrew/.linuxbrew/opt/libtool/lib:/home/linuxbrew/.linuxbrew/opt/zlib/lib:/home/linuxbrew/.linuxbrew/opt/openssl/lib
```

`zlib` and `openssl` make sense, but I don't think `libtool` should be there.

Putting some debugging code in revealed `autoconf` and `automake`, both build deps, were also considered, but they didn't end up in the RPATH presumably because they don't have lib directories.

sjackman replied:
:+1: Nice catch. Fixed.

sjackman replied:
I haven't yet fixed this for the `rpath` PR for `--env=std`. There's shared code, so I think I'll merge this PR first, and then rebase the `rpath` code on top of this merged PR.

bw-yuhogg commented on Library/Homebrew/build.rb:
```diff
@@ -81,6 +82,7 @@ def install
       ENV.keg_only_deps = keg_only_deps
       ENV.deps = formula_deps
+      ENV.run_time_deps = run_time_deps
       ENV.x11 = reqs.any? { |rq| rq.is_a?(X11Requirement) }
```
Tangent: Does this need to be changed to also support XorgRequirement?

Note that when we port that to Homebrew/brew, I think we'll probably drop that name and instead use the extension mechanism Mike made, which is going to require some rather substantive retesting on our end come merge time.

sjackman replied:
It's used to point to paths to in the XQuartz SDK, which shouldn't be necessary for `xorg` because it's in the brew prefix.

sjackman committed:
superenv: Remove macOS-specific code for Linuxbrew

sjackman committed:
superenv: Fix keep? for Linuxbrew

sjackman committed:
super/cc: Remove redundant RPATH for Linuxbrew

Using -rpath=/a -rpath=/a:/b results in /a:/a:/b whereas
using -rpath=/a -rpath=/a -rpath=/b results in /a:/b

maxim-belkin commented on Library/Homebrew/shims/linux/super/cc:
```diff
@@ -0,0 +1,346 @@
+  exit 1
+fi
+exec "$HOMEBREW_RUBY_PATH" -x "$0" "$@"
+#!/usr/bin/env ruby -W0
```
Interesting approach... :)
